### ISSUE TITLE
4.14.0 Release checklist

### ISSUE THREAD

**Issue by compulim** [microsoft/BotFramework-WebChat#3981]
# Checklist

## Build

1. [x] Bump MockBot to Bot Framework SDK release 4.14.0
   - https://www.npmjs.com/package/botbuilder/v/4.14.0
1. [x] ~Bump `botframework-directlinejs` to `0.14.1`~
   - No newer version
1. [x] Bump to `4.14.0`
   - [x] Update `CHANGELOG.md` to mark specific changes in `4.14.0`
   - [x] Run `npm version --no-git-tag-version 4.14.0`
   - [x] Merged into `master`, the PR number is #3984
   - Commit is `af70aaa`
   - Do not merge any other unrelated changes after this PR. Any other PR merged, will need to be re-tested
1. [x] Run daily pipeline manually, set "generate release version number" to `true`
   - (This will not push to NPM or CDN)
   - Pipeline name is `BotFramework-WebChat-daily`
   - The build number is [`257964`](https://fuselabs.visualstudio.com/BotFramework-WebChat/_build/results?buildId=257964&view=results) and commit is `af70aaa`
1. [x] Wait for [`WebChat-release-testing` pipeline](https://fuselabs.visualstudio.com/BotFramework-WebChat/_release?_a=releases&view=mine&definitionId=17) to complete
   - Pipeline name is `Push-Release-Testing-to-GitHub-Pages`
   - The release ID is [`2751`](https://fuselabs.visualstudio.com/BotFramework-WebChat/_releaseProgress?_a=release-pipeline-progress&releaseId=2751)
1. [x] Add manual tests to WebChat-release-testing as needed

## Test

> The test should run against the build artifacts from Azure Pipelines.

1. [x] Manual testing on major browsers using `webchat-release-testing`
   - [x] Before starting testing, update all the browser version to latest
   - [x] Chrome 91.0.4472.124
   - [x] Edge 91.0.864.64
   - [x] Firefox 89.0.2
   - [x] IE11 (IE Mode under Windows 11 22000.51)
   - [x] macOS Safari 14.1.1 (16611.2.7.1.4)
   - [x] iOS Safari 14.6
   - [x] Android Chrome 91.0.4472.120
1. [x] Test specific fixes related to `4.14.0` and previous releases
   
   - Using a new engine to strip Markdown (PR [#3917](https://github.com/microsoft/BotFramework-WebChat/pull/3917))

## Release

1. [x] Verify on WebChat-release-testing
1. [x] Make sure you are on `master` ~or `qfe`~ branch, run `git status` to check
1. [x] `git pull`
1. [x] Verify `/package.json`, `/package-lock.json`, and `CHANGELOG.md` has a version of `4.14.0`
1. [x] `git log`
   - Verify the latest commit is `af70aaa`
1. [x] `git tag v4.14.0`
1. [x] `git push -u upstream v4.14.0`
   - You do not need to kick off a build again, use the previous build
1. [x] Create a new GitHub release, copy entries from `CHANGELOG.md`
   - [x] Subresource Integrity can be generated by
      - From CDN: `curl -H 'Accept-Encoding: gzip' https://cdn.botframework.com/botframework-webchat/4.14.0/webchat.js | gunzip - | openssl dgst -sha384 -binary | openssl base64 -A`
      - From local: `cat webchat.js | openssl dgst -sha384 -binary | openssl base64 -A`
   - [x] Attach assets including 3 JS files, `stats.json` and 5 tarballs
      - You can copy the artifacts from `webchat-release-testing/drops`
      - Tarballs download from npmjs
         ```
         curl -L -o botframework-directlinespeech-sdk-4.14.0.tgz https://registry.npmjs.org/botframework-directlinespeech-sdk/-/botframework-directlinespeech-sdk-4.14.0.tgz
         curl -L -o botframework-webchat-4.14.0.tgz https://registry.npmjs.org/botframework-webchat/-/botframework-webchat-4.14.0.tgz
         curl -L -o botframework-webchat-core-4.14.0.tgz https://registry.npmjs.org/botframework-webchat-core/-/botframework-webchat-core-4.14.0.tgz
         curl -L -o botframework-webchat-api-4.14.0.tgz https://registry.npmjs.org/botframework-webchat-api/-/botframework-webchat-api-4.14.0.tgz
         curl -L -o botframework-webchat-component-4.14.0.tgz https://registry.npmjs.org/botframework-webchat-component/-/botframework-webchat-component-4.14.0.tgz
         ```
1. [x] Kick off release to NPM
   - Release name is `[[PROD]]Push-WebChat-to-npmjs`
   - The build number is [257964](https://fuselabs.visualstudio.com/BotFramework-WebChat/_build/results?buildId=257964&view=results) release number is [`33`](https://fuselabs.visualstudio.com/BotFramework-WebChat/_releaseProgress?_a=release-pipeline-progress&releaseId=2752) and commit is `af70aaa`
   - [x] Verify package content then click Resume
   - [x] Retain the build indefinitely
1. [x] Kick off release to CDN (CDN CUTOFF TIME IS 2PM PST)
   1. [x] Prepare the email for approval
   - Release name is `[[PROD]]Push-WebChat-to-Prod-CDN-with-approval`
   - The build number is [`257964`](https://fuselabs.visualstudio.com/BotFramework-WebChat/_build/results?buildId=257964&view=results), release number is [`39`](https://fuselabs.visualstudio.com/BotFramework-WebChat/_releaseProgress?_a=release-environment-logs&releaseId=2755&environmentId=2755) and commit is `af70aaa`
   - Script build number is `131156` (this is fixed)
   1. [x] Send reminder email to approvers
   - Retain the build indefinitely

## Post-release verification - complete within 30mins / 1hr of the release

- [x] Test using `webchat-release-testing`
   1. [x] Clone https://github.com/corinagum/WebChat-release-testing/
   1. [x] `01.create-react-app`
      1. [x] Nuke `01.create-react-app/node_modules`
      1. [x] `npm install`
      1. [x] `npm install botframework-webchat@4.14.0` (just install the bundle package)
      1. [x] `npm run build`
   1. [x] Others
      -  Using script tags from https://github.com/microsoft/BotFramework-WebChat/releases/tag/v4.14.0, with subresource integrity
         ```
         <script
            crossorigin="anonymous"
            integrity="sha384-"
            src="https://cdn.botframework.com/botframework-webchat/latest/webchat-es5.js"
         ></script>
         ```
   1. [x] `npx serve` (at repo root)
   1. [x] Go to http://localhost:5000/ to test, including IE11

## Notification to interested parties

- [x] Update partner page on Adaptive Cards doc
   - https://docs.microsoft.com/en-us/adaptive-cards/resources/partners
   - PR at https://github.com/MicrosoftDocs/AdaptiveCards/pull/405
- [ ] Notify related parties for the following fixes
   - [x] SDK team
   - [ ] Omnichannel
   - [ ] Pooja
- [ ] Update root README.md with feature notes -- Note: PR will be combined with post-release checklist PR
- [x] Update https://github.com/microsoft/BotFramework-sdk/issues/6334
   
---

# Post-release checklist

These are chores that we should do before starting the cycle to reduce ripple effects if we do it in mid-cycle.

Tips: 

- Clean your repo before start
- Remove `node_modules` from all folder
   - `git clean -fdx`
- Never delete `package-lock.json`
- If you mess it up, tableflip and redo
- In `component/package.json`
   - Remove reference to `botframework-webchat-core` by hand-modifying `package.json`
   - Then, `npm install` (symlinks will be broken afterward)
   - Then, add those references back by hand-modifying `package.json`
   - This also applies for other packages with similar dependencies/symlinks
   - To build afterward, do tableflip to rebuild those symlinks

## Applies to all releases

> This list should be copied to versions in the future.

- [x] ~If on QFE branch, make sure `CHANGELOG.md` and version number bump is cherry-picked to `master`~
   - [ ] ~`git checkout master`~
   - [ ] ~`git cherry-pick XXX` (the commitish for bumping version number and `CHANGELOG.md`)~
- [x] Correct the date for 4.14.0 in `CHANGELOG.md`
   - There could be last minute fixes that could push the planned date later than the one in `CHANGELOG.md`
- [x] Bump `package.json` to `4.14.1-0`
   - Run `npm version prepatch --no-git-tag-version`
- [x] Update [`servicingPlan.json`](https://github.com/microsoft/BotFramework-WebChat/blob/master/packages/embed/servicingPlan.json)
   - Add deprecation notes for previous versions
   - Subresource integrity hash from https://github.com/microsoft/BotFramework-WebChat/releases/tag/v4.14.0
- [x] Update all samples to use `4.14.0`
   - Some samples are pointing to GitHub Releases because the sample need new features from daily build
   - Search "https://github.com/microsoft/BotFramework-WebChat/releases/download/"
      - And replace with "https://cdn.botframework.com/botframework-webchat/latest/"
- [ ] Clean up unnecessary branch on official repo

The PR is https://github.com/microsoft/BotFramework-WebChat/pull/XXX

### Bump all dependencies to latest version

On each `package.json` under root and `/packages/`:

- [ ] All NPM packages (`@babel/*`, `jest`, `lerna`, `typescript`, `webpack`, etc) in PR #XXX
   - Don't bump the following packages
      - `react`, `react-dom`, `redux`, `redux-saga`, `@types/react`
      - `sanitize-html`
         - When bumping `sanitize-html`, make sure `bundle/postinstall` scripts continue to works as we only transpile a single `index.js` file and assume the `sanitize-html` package will keep it that way
      - `serve`
         - Because newer version could be buggy on Windows
      - `rxjs@5` on `playground`
      - `webpack@4`
         - Starting from version 5, they dropped Node.js polyfills for browsers
         - Keep the following dependencies because their latest version requires `webpack@5`
            - [`html-webpack-plugin@4`](https://npmjs.com/package/html-webpack-plugin)
            - [`source-map-loader`](https://npmjs.com/package/source-map-loader)
      - `microsoft-cognitiveservices-speech-sdk`
        - We will bump it in a separate PR
      - `adaptivecards`
        - We will bump it in a separate PR
      - `expect@25` under `test/harness`
        - The newer version requires Node.js (`graceful-fs`)
- [ ] Run `npm audit fix` to make sure everything is fixed
- [ ] Test under IE11 to make sure all dependencies are working
- [ ] List steps to verify bumping `microsoft-cognitiveservices-speech-sdk`

#### Obstacles to bump `npm`

As of writing, `npm@7.19.1` (bundled by `node@16`) has issue on running `postinstall` script.

Also, `microsoft-cognitiveservices-speech-sdk@1.17.0` is peer-depends on `@angular/common`, which `npm@7` will automatically install missing peer dependencies. [Track the issue here](https://github.com/microsoft/cognitive-services-speech-sdk-js/issues/403).

We need at least both issues to be fixed before we can jump to `npm@7`.

### Bump Docker image

> The Docker image can be found at root `docker-compose.yml` and `Dockerfile*`.

- [ ] Docker container for headless Chrome (#XXX)
   - They recently moved from `3.14.159-xxx` tag scheme to a more sensible `87.0` tag scheme
   - Tags can be found at https://hub.docker.com/r/selenium/node-chrome/tags
   - Preferably in separate PR because screenshots change can be large occasionally
   - Run tests locally, as the screenshots can be slightly different

**Comment by christopherjbaker** [comment]
I would like to request that you update `sanitize-html` whether or not in works with IE11. I'm sure you can figure out a polyfil that makes those things work, but even without that. Versions prior to 2.0, as well as some of the 2+ ones, have [_major_ security issues](https://snyk.io/vuln/search?type=npm&q=sanitize-html), and holding those back for a browser that nobody uses (see image) seems like a very, very bad idea. 
![image](https://user-images.githubusercontent.com/980338/125864096-75b8dd3d-0ef8-4ef9-8208-3ed60d853692.png)

**Comment by compulim** [comment]
Will see what I can do with `sanitize-html`. I am still working on bumping dependencies.

Security issues and platform support are two different things. We will do both.

In my experience, issues due to IE11 are much easier/shorter to resolve than one of the green bars in your chart.

**Comment by christopherjbaker** [comment]
Ok, thanks! I'll check back in later.

I did do some quick looking over there at what isn't IE11 compatible. They direct people to [this section](https://github.com/apostrophecms/sanitize-html#browser) of their readme. They changed the paradigm of their project, and no longer provide a browser-ready build, instead telling dependants to do the build themselves if needed. I imagine most projects using you on the FE will do so as part of a build, and it looks like your project also has a build (I didn't do the botframework integration, so I don't know much about this project), so as long as those builds add IE11 pollyfils when they need compatibility, there shouldn't be anything to do to upgrade (from a quick look, it doesn't look like their API changed between 1 and 2, just that paradigm). Hopefully this helps.

**Comment by compulim** [comment]
Thanks for the tips!

Yap, polyfill + transpile `sanitize-html` should work for IE11. That's what we are doing from time to time too. And that's why I say fixing IE11 issues are, most of the time, easier than fixing another browser.

But sometimes, transpiling could be tricky. Let me share all my pain points and experiences to you.

Web Chat have 2 deliverables:

- Bundle (`*.js`), it's transpiled through Webpack, we can configure it easily
- NPM packages, it will be up to web devs on how to transpile it
   - If they are using `create-react-app` (or anything that transpile `/node modules/`), they should be fine
   - If they build their own pipeline, it may not easy for them to configure the pipeline
      - The team which want to integrate our component, may not be the same team which own the build pipeline. Thus, they may not be authorized to configure it

Sometimes, architecting web components could be more complex than web apps. We need to think about the customer's team structure, how they are going to write/build/architect their apps, how they are integrating our component.

For transpiling, one trick is to use `postinstall` script to transpile packages on-the-fly. Then a few issues along this route:

- Recently `npm@7` has a bug on `postinstall` and it still repro on `7.19.1` as of yesterday
   - NPM team just released `7.20.0` hours ago, I believe it is not fixed yet, but I should give it a try
- If custom NPM registry is configured with CI/CD pipelines in Azure DevOps, `npx` during `postinstall` will fail 😭
   - Workaround: we need to install `babel` as a production dependency, instead of relying on `npx` or as dev dependencies... not great, but not too bad

Anyway... I will find ways to transpile `sanitize-html` on-the-fly. Many options in my head, need to find the leanest way to do the job.

BTW, thanks for choosing our component.

**Comment by compulim** [comment]
FYI, I tested transpiling `sanitize-html@2.4.0` and it works on IE11. 😉

**Comment by compulim** [comment]
@christopherjbaker 4.14.1 is out. `sanitize-html` is bumped to `2.4.0`. Now with zero production-hitting vulnerabilities (as of now). Also works with IE11. Everyone should be happy. 😉

![image](https://user-images.githubusercontent.com/1622400/132592730-32fb4c56-f8f2-45f1-a93d-d26191f95434.png)

FYI, `sanitize-html@2.5.0` was released just an hour ago.
